pkg/network/server.go

```diff
@@ -274,20 +278,72 @@ func (s *Server) Shutdown() {
 
 // AddService allows to add a service to be started/stopped by Server.
 func (s *Server) AddService(svc Service) {
+	s.serviceLock.Lock()
+	defer s.serviceLock.Unlock()
+	s.addService(svc)
+}
+
+// addService is an unlocked version of AddService.
+func (s *Server) addService(svc Service) {
 	s.services[svc.Name()] = svc
 }
 
 // AddExtensibleService register a service that handles an extensible payload of some kind.
 func (s *Server) AddExtensibleService(svc Service, category string, handler func(*payload.Extensible) error) {
+	s.serviceLock.Lock()
+	defer s.serviceLock.Unlock()
+	s.addExtensibleService(svc, category, handler)
+}
+
+// addExtensibleService is an unlocked version of addExtensibleService.
+func (s *Server) addExtensibleService(svc Service, category string, handler func(*payload.Extensible) error) {
 	s.extensHandlers[category] = handler
-	s.AddService(svc)
+	s.addService(svc)
 }
 
 // AddExtensibleHPService registers a high-priority service that handles an extensible payload of some kind.
 func (s *Server) AddExtensibleHPService(svc Service, category string, handler func(*payload.Extensible) error, txCallback func(*transaction.Transaction)) {
+	s.serviceLock.Lock()
+	defer s.serviceLock.Unlock()
 	s.txCallback = txCallback
 	s.extensHighPrio = category
-	s.AddExtensibleService(svc, category, handler)
+	s.addExtensibleService(svc, category, handler)
+}
+
+// DelService drops a service from the list, use it when the service is stopped
+// outside of the Server.
+func (s *Server) DelService(svc Service) {
+	s.serviceLock.Lock()
+	defer s.serviceLock.Unlock()
+	s.delService(svc)
+}
+
+// delService is an unlocked version of DelService.
+func (s *Server) delService(svc Service) {
+	delete(s.services, svc.Name())
+}
+
+// DelExtensibleService drops a service that handler extensible payloads from the
+// list, use it when the service is stopped outside of the Server.
+func (s *Server) DelExtensibleService(svc Service, category string) {
+	s.serviceLock.Lock()
+	defer s.serviceLock.Unlock()
+	s.delExtensibleService(svc, category)
+}
+
+// delExtensibleService is an unlocked version of DelExtensibleService.
+func (s *Server) delExtensibleService(svc Service, category string) {
+	delete(s.extensHandlers, category)
+	s.delService(svc)
+}
+
+// DelExtensibleHPService unregisters a high-priority service that handles an extensible payload of some kind.
+func (s *Server) DelExtensibleHPService(svc Service, category string) {
+	s.serviceLock.Lock()
+	defer s.serviceLock.Unlock()
+	s.txCallback = nil
+	s.extensHighPrio = ""
```

Comment:
Why did you decide to stop processing high-priority payloads as such, even if the service itself is disabled?

Reply:
That's just symmetry, although yes, it's controversial. At the same time on non-CN nodes we don't have high priorities at all now which probably isn't good. Need to think about it.

Reply:
Fixed.